tools/common-subexpr-elim.rkt

```diff
@@ -0,0 +1,67 @@
+#lang racket
+
+(require "common.rkt" "fpcore.rkt")
+
+(provide core-common-subexpr-elim)
+
+(define (core-common-subexpr-elim prog)
+  (match-define (list 'FPCore (list args ...) props ... body) prog)
+  (define cs-hash (count-common-subexpr body))
+  (define cse-body (common-subexpr-elim cs-hash body))
+  `(FPCore ,args ,@props ,cse-body))
+
+(define (combine-common-subexpr-hash h1 h2)
+  (for ([(k v) (in-hash h2)])
+    (if (hash-has-key? h1 k)
+      (hash-set! h1 k #t)
+      (hash-set! h1 k v)))
+  h1)
+
+(define (count-common-subexpr expr)
+  (cond 
+    [(list? expr)
```

Comment:
I think this works for function calls, but for `let`s and `while`s:

1. It will try to eliminate common variable bindings, which is incorrect since they're not expressions.
2. It will skip the first variable binding since it thinks it is an operator
3. It will misbehave in the presence of shadowing, which is sadly quite common in things compile from FPImp.

Reply:
An "easy" way to deal with this is to stop recursing at `let`s and `while`s. That's not ideal though—you could eliminate more expressions if you recurse, but filter out all expressions with bound variables.